### DISCUSSION TITLE
tar_map_rep - support for nested futures?

### DISCUSSION THREAD

**Discussion by kkmann** [ropensci/tarchetypes#117] [Help] [answered]
Hi,

I think the example in [tar_map-rep()](https://docs.ropensci.org/tarchetypes/reference/tar_map_rep.html) pretty much perfectly matches my use case: Simulation of operating characteristics over a static table of parameter values.

What's unclear to me is how `tar_map_rep()` interacts with `tar_make_future` and a nested execution plan like

```
 plan(list(
    future.batchtools::batchtools_slurm, 
    multicore
  ))
```

I would like the "map" part to be dispatched to individual compute nodes and the "rep" part to be handled by multicore parallellism within each compute node. This would allow me to send each "scenario" to a separate worker on the slurm cluster and then make use of efficient multicore parallelism to resample the scenario on the worker node. 

So far, I have been using dynamic branching with a `pattern` statement and manually implemented the resampling within the simulation function using `furrr::map`. When executing the pipeline with `tar_make_future` that produces the desired nested futures. This has two disadvantages. 1) I need to wrap the actual simulation function in furrr::map 2) I have no control over the chunk size of scenarios being bundled up for a single worker node. If an individual scenario takes little time but there are many scenarios, this leads to a lot of overhead by sending small jobs on the cluster. 

Is there currently a way within targets/tarchetypes to get `tar_map_rep` to work with nested futures (slurm/multicore) and control scheduling to the cluster at the same time?

**Comment by wlandau** [comment, marked as answer]
That's a great idea, very fitting for batched replication. Implemented in https://github.com/ropensci/tarchetypes/commit/14ece41813e977aa096147f08e0bdfc239038987.

NB as you saw in https://github.com/HenrikBengtsson/future.batchtools/issues/65, `tar_make_future()` suffers from `batchtools` overhead with e.g. SLURM. My group and I have had better luck with `tar_make_clustermq()` with [these performance tweaks](https://books.ropensci.org/targets/performance.html). But strict walltime caps might get in the way of persistent workers.

**Reply by kkmann** [reply]
I'll take it for a spin :) I see the advantage of persistent clusters, but slurm has its advantages... This is pretty much why I was stuck using nested `future_map()` calls so far. The overhead can be reduced by controlling the batch size in the outer loop. In my experience, anything around 50-100 slurm jobs is fine, it gets difficult with more than 200 jobs depending on the HDD/network speed.

**Reply by wlandau** [reply]
@kkmann, in the GitHub version of `tarchetypes`, I ended up needing to disable nested parallelism by default. ~~However, there is a new `parallel_reps` argument in `tar_map_rep()` etc. that will let you reactivate it.~~

**Reply by wlandau** [reply]
On reflection, I changed within-batched parallelism to multiprocess only using `future.callr::callr`. This makes things consistent across `tar_make_future()`, `tar_make_clustermq()`, and the eventual version that uses [`crew`](https://github.com/wlandau/crew). `parallel_reps` is now `rep_workers` and sets the number of local parallel processes.
